src/node_internals.h

```diff
@@ -297,6 +299,91 @@ v8::MaybeLocal<v8::Value> StartExecution(Environment* env,
 namespace coverage {
 bool StartCoverageCollection(Environment* env);
 }
+
+#ifdef _WIN32
+typedef SYSTEMTIME TIME_TYPE;
+#define PATHSEP "\\"
+#else  // UNIX, OSX
+typedef struct tm TIME_TYPE;
+#define PATHSEP "/"
+#endif
+
+static void LocalTime(TIME_TYPE* tm_struct) {
+#ifdef _WIN32
+  GetLocalTime(tm_struct);
+#else  // UNIX, OSX
+  struct timeval time_val;
+  gettimeofday(&time_val, nullptr);
+  localtime_r(&time_val.tv_sec, tm_struct);
+#endif
+}
+
+class DiagnosticFilename {
+ public:
+  DiagnosticFilename(Environment* env,
+                     const char* prefix,
+                     const char* ext,
+                     int seq = -1) :
+      filename_(MakeFilename(env->thread_id(), prefix, ext, seq)) {}
+
+  DiagnosticFilename(uint64_t thread_id,
+                     const char* prefix,
+                     const char* ext,
+                     int seq = -1) :
+      filename_(MakeFilename(thread_id, prefix, ext, seq)) {}
+
+  const char* operator*() const { return filename_.c_str(); }
+
+ private:
+  static const std::string& MakeFilename(
```

Comment:
```suggestion
  static std::string MakeFilename(
```

(Returning a reference to a local variable is generally a bug)

I’d also maybe put this function in a separate source file, it’s kind of large and doesn’t seem like it should be inlined + duplicated each time we use it.